libs/linearSolver/linearSolverPMINRES.cpp

```diff
@@ -0,0 +1,156 @@
+                               "innerProd", "weightedInnerProd",
+                               "norm2", "weightedNorm2"});
+
+  dlong Ntotal = N + Nhalo;
+
+  dfloat *dummy = new dfloat[Ntotal]();
+  o_p     = platform.malloc(Ntotal*sizeof(dfloat), dummy);
+  o_z     = platform.malloc(Ntotal*sizeof(dfloat), dummy);
+  o_r     = platform.malloc(Ntotal*sizeof(dfloat), dummy);
+  o_r_old = platform.malloc(Ntotal*sizeof(dfloat), dummy);
+  o_q     = platform.malloc(Ntotal*sizeof(dfloat), dummy);
+  o_q_old = platform.malloc(Ntotal*sizeof(dfloat), dummy);
+  delete[] dummy;
+
+  weighted = _weighted;
+  o_w = _o_weight;
+
+  occa::properties kernelInfo = platform.props;
+  updateMINRESKernel = platform.buildKernel(LINEARSOLVER_DIR "/okl/linearSolverUpdateMINRES.okl", "updateMINRES", kernelInfo);
+
+  return;
+}
+
+pminres::~pminres()
+{
+  return;
+}
+
+int pminres::Solve(solver_t& solver, precon_t& precon,
+                   occa::memory &o_x, occa::memory &o_b,
+                   const dfloat tol, const int MAXIT, const int verbose)
+{
+  int    rank, iter;
+  dfloat a0, a1, a2, a3, del, gam, gamp, c, cp, s, sp, eta;
+  dfloat TOL;
+
+  MPI_Comm_rank(comm, &rank);
+  linAlg_t &linAlg = platform.linAlg;
+
+  solver.Operator(o_x, o_r);            // r = b - A*x
+  linAlg.axpy(N, 1.0, o_b, -1.0, o_r);
+  precon.Operator(o_r, o_z);            // z = M\r
+
+  gamp = 0.0;
+  gam  = sqrt(innerProd(o_z, o_r));     // gam = sqrt(z . r);
+  eta  = gam;
+  sp   = 0.0;
+  s    = 0.0;
+  cp   = 1.0;
+  c    = 1.0;
+
+  TOL = mymax(tol*fabs(eta), tol);
+  if (verbose && (rank == 0)) {
+    printf("PMINRES:  initial eta = % .15e, target %.15e\n", eta, tol);
+  }
+
+  // MINRES iteration loop.
+  iter = 0;
+  while (iter < MAXIT) {
+    if (verbose && (rank == 0)) {
+      printf("PMINRES:  it %3d  eta = % .15e, gamma = %.15e\n", iter, eta, gam);
+    }
+
+    if ((fabs(eta) < TOL) && (iter >= 1)) {
+      if (verbose && (rank == 0)) {
+        printf("PMINRES converged in %d iterations (eta = % .15e).\n", iter, eta);
+        return iter;
+      }
+    }
+
+    linAlg.scale(N, 1.0/gam, o_z);                    // z = z/gam
+    solver.Operator(o_z, o_p);                        // p = A*z
+    del = innerProd(o_p, o_z);                        // del = p . z
+    a0 = c*del - cp*s*gam;
+    a2 = s*del + cp*c*gam;
+    a3 = sp*gam;
+#if 0
+    linAlg.axpy(N, -a2, o_q, 1.0, o_z);               // z = z - a2*q - a3*q_old
+    linAlg.axpy(N, -a3, o_q_old, 1.0, o_z);
+    o_q_old.copyFrom(o_q);                            // q_old = q
+    o_q.copyFrom(o_z);                                // q = z
+    o_z.copyFrom(o_r);                                // z = r
+    linAlg.axpy(N, 1.0, o_p, -(del/gam), o_r);        // r = p - (del/gam)*r
+    if (iter > 0)
+      linAlg.axpy(N, -(gam/gamp), o_r_old, 1.0, o_r); // r = r - (gam/gamp)*r_old
+    o_r_old.copyFrom(o_z);                            // r_old = z
+#else
```

Comment:
Debug code, or just an initial implementation? You want to keep this?

Reply:
Yes, it's an initial implementation, but I do want to keep it around as a reference for the time being.